manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
+
+\subsection{s:mm_semantics_reasoning}{Reasoning with the semantics}
+
+Let us revisit an example from earlier (section \ref{s:why_relaxed_memory}).
+
+\begin{caml_example*}{verbatim}
+let a = ref 0
+and b = ref 0
+
+let d1 () =
+  a := 1;
+  !b
+
+let d2 () =
+  b := 1;
+  !a
+
+let main () =
+  let h = Domain.spawn d2 in
+  let r1 = d1 () in
+  let r2 = Domain.join h in
+  assert (not (r1 = 0 && r2 = 0))
+\end{caml_example*}
+
+This program has a data race on "a" and "b", and hence, the program may exhibit
+non sequentially consistent behaviour. Let us use the semantics to show that
+the program may exhibit "r1 = 0 && r2 = 0".
+
+The initial state of the abstract machine is:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t0 -> 0]
+b: [t1 -> 0]
+
+(* Domains *)
+d1: [a -> t0; b -> t1]
+d2: [a -> t0; b -> t1]
+\end{verbatim}
+
+There are several possible schedules for executing this program. Let us
+consider the following schedule:
+
+\begin{verbatim}
+1: a := 1 @ d1
+2: b := 1 @ d2
+3: !b     @ d1
+4: !a     @ d2
+\end{verbatim}
+
+After the first action "a:=1" by "d1", the machine state is:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t0 -> 0; t2 -> 1] (* new write at t2 *)
+b: [t1 -> 0]
+
+(* Domains *)
+d1: [a -> t2; b -> t1] (* frontier updated at a *)
+d2: [a -> t0; b -> t1]
+\end{verbatim}
+
+After the second action "b:=1" by "d2", the machine state is:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t0 -> 0; t2 -> 1]
+b: [t1 -> 0; t3 -> 1] (* new write at t3 *)
+
+(* Domains *)
+d1: [a -> t2; b -> t1]
+d2: [a -> t0; b -> t3] (* frontier updated at b *)
+\end{verbatim}
+
+Now, for the third action "!b" by "d1", observe that the "d1"'s frontier at "b"
+is at "t1". Hence, the read may return either "0" or "1". Let us assume that it
+returns "0". The machine state is not updated by the non-atomic read.
+
+Similarly, for the forth action "!a" by "d2", "d2"'s frontier at "a" is at
+"t0". Hence, this read may also return either "0" or "1". Let us assume that it
+returns "0". Hence, the assertion in the original program, "assert (not (r1 = 0
+&& r2 = 0))", will fail for this particular execution.
+
+\section{s:mm_tearing}{Non-compliant operations}
+
+There are certain operations which are not memory model compliant.
+
+\begin{itemize}
+  \item "Array.blit" function on float arrays may cause \emph{tearing}. When an
+    unsynchronized blit operation runs concurrently with some overlapping write
+    to the fields of the some float array, the field may end up with bits from
+    either of the writes.
+  \item With flat-float arrays on 32-bit architectures, getting or setting a
+    field involves two separate memory accesses. In the presence of data races,
+    the user may observe tearing.
+  \item The "Bytes" module~\stdmoduleref{Bytes} permits mixed-mode accesses
+    where reads and writes may be of different sizes. Unsynchronized mixed-mode
+    accesses lead to tearing.
```

Comment:
Maybe we should document the fact that some `Bytes` functions step out of the memory model. E.g.:
```
    \item Unsynchronized accesses to the same sub-range of "bytes" using
      functions from the \stdmoduleref{Bytes} module, such as "Bytes.blit" or
      "Bytes.fill", are an exception to the memory model. Concurrent writes or
      concurrent reads and writes to the same bytes invalidate any guarantees
      about the program. This warning extends to all standard functions that
      writes to "bytes" directly, such as "Unix.recv".
```

Reply:
First, the problem is about accesses to bytes of the same byte array, not on ly to the same bytes. We could even say that there is a problem when accessing to bytes in the same 64bits (or 32 bits) block in a byte array only.

Second, we should take care about bounding the undefinedness to the array of bytes. This is the general spirit of the memory model that a race condition is bounded in space and time. I.e., we could say that all bets are off when it comes to values read from that bytes array until the next synchronized write.

Reply:
Or we could use in C relaxed atomic accesses that do not suffer from UB, but do not incur significant overheads since they are compiled to plain accesses.

I think that breaking type safety because of this is not an option.